### ISSUE TITLE
Code fix for CA1826 "Do not use Enumerable methods on indexable collections. Instead use the collection directly"

### ISSUE THREAD

**Issue by Neme12** [dotnet/roslyn-analyzers#1932]
It would be nice to have a code fix at least for the case of `First()` and `Last()`.

**Comment by TKharaishvili** [comment]
Is this issue labeled correctly?

I just installed `Microsoft.CodeQuality.Analyzers` in the .NET 4.7 console app and couldn't find `CA1826` anywhere in the analyzer list.

After searching around some more I found it in the `Microsoft.NetCore.Analyzers` package:

![image](https://user-images.githubusercontent.com/4363170/53642249-c4e37c00-3c4a-11e9-87d8-2c9911058bdc.png)

but writing this piece of code gives me no warnings:
```C#
var list = new List<string> { "a", "b", "c" };
var f = list.First();
var l = list.Last();
var count = **list.Count();
```
since it says `NetCore.Analyzers` I then tried it on a .net core console app but still got no warnings.

Needless to say I'm quite confused.

**Comment by TKharaishvili** [comment]
Just looked at the source code and found out that the warning is not issued for `IList<T>`-s. Probably because those enumerable methods are optimized for them.

But the question of this issue not being labeled correctly still remains.

**Comment by TKharaishvili** [comment]
Hi @mavasani I just created a PR for this - https://github.com/dotnet/roslyn-analyzers/pull/2173

It's not ideal but I'm hoping you can guide me to the right direction with some code reviews. Cheers.

**Comment by Evangelink** [comment]
A code fix now exists for `First`, `Last` and `Count` so I think the intent of this issue is actually solved.

The remaining work is:

- Handle `FirstOrDefault` and `LastOrDefault` 

- Handle C# 8 index expression for the codefix of `Last` and `LastOrDefault`. This is optional as some other codefix could help the migration to this new syntax

- Handle the creation on an intermediate variable for `Last` and `LastOrDefault`. This case can be a bit tricky and lead to various modifications in the case of an arrow expression or an invocation.

**Comment by paul1956** [comment]
This warning should not apply to VB or at least there should be a Fix to locally suppress it and I don't see what the fix would be for the OrDefault versions without the code getting much more complex.

**Comment by Evangelink** [comment]
@paul1956 Could you be a little more specific, are you talking about the rule in general? Are you referring only to `Last` and `LastOrDefault`? I am not experienced with vbnet but this does seem or for `First`, `FirstOrDefault` and `Count`. If you disagree could you add some examples to better explain your point?

**Comment by paul1956** [comment]
@Evangelink Using the collection directly in VB is not easy and makes the code very complex and there is no CodeFix for VB. Where List is some collection:
First and maybe Count when Length is available is easy
List.First vs List(0) is a wash as is List.Count vs List.Length
VB does not have Index expressions, there is no easy way to do FirstOrDefault or LastOrDefault.
Last is one that could be debated but Last seems simpler
List.Last vs List(List.Count-1)

Generally It would be better is this was just disabled for VB

**Comment by CodingOctocat** [comment]
The analyzer should exclude the `OrDefault()` method. It's annoying and I need more code if I don't use `FirstOrDefault`.

```cs
// Models is IReadOnlyList
public Model? Model=> QueryResponse?.Models?.FirstOrDefault(); // CA1826
```

**Comment by trivalik** [comment]
@CodingOctocat I think `FirstOrDefault` is just the case if a predicate is used.

Slower:
```
                    var list = new List<int> { 1, 2, 3, 4, 5, 6 };
                    var firstOrDefault = list.FirstOrDefault(x => x == 5);
```

Faster:
```
                    var list = new List<int> { 1, 2, 3, 4, 5, 6 };
                    var firstOrDefault = list.Find(x => x == 5);
```

**Comment by MohammedZakirMemon** [comment]
Hi, I’d like to work on one of the remaining parts of this issue. I’m planning to investigate adding C# 8 index expression support for the existing Last() code fix, so it can use [^1] where appropriate.

**Comment by kharai-tor** [comment]
@MohammedZakirMemon They've moved analyzers from this repo to other repos. Have a look at the readme here - https://github.com/dotnet/roslyn-analyzers
